### ISSUE TITLE
propagation of attributes in RealField/ComplexField

### ISSUE THREAD

**Issue by nickhand** [rainwoodman/pmesh#12]
the `pm` attributes aren't attached to the return Field when using numpy functions, i.e., 

``` python
real = RealField(pm)
real2 = numpy.exp(real)
real2.pm # raises AttributeError
```

I think we just need to add an ``__array_finalize__(self, obj)`` function to `Field` that attaches the relevant attributes. Does that make sense?

**Comment by nickhand** [comment]
as described here: https://docs.scipy.org/doc/numpy/user/basics.subclassing.html

**Comment by rainwoodman** [comment]
Alternatively, we can assert any ufunc return value is cast to a regular numpy array (sub-class removed), and make the cast of subclass explicity with numpy.view(). 

If this is doable it tends to produce cleaner code than relying on the messy array finalize stuff in numpy, I feel.

BTW, this case can be replaced with a inplace operation:
```
numpy.exp(real, real)
```
or fake in-place operation,
```
real[...] = numpy.exp(real)
```

**Comment by nickhand** [comment]
In my case, I wanted to make a copy, so I had to first create the object with the constructor and then do the filling operation. 

I think I would side with attaching the relevant attributes in an array finalize, but I don't feel strongly either way. Right now, the code is very confusing though. Even doing something like this breaks unexpectedly

```python
 pm = ParticleMesh(BoxSize=512, Nmesh=[256]*3, dtype='f4')
real = RealField(pm)
real2 = real*0.
pm2 = real2.pm # raises AttributeError
```

I guess I had written some code assuming it behaves like a numpy array and then had to trace back the origin of the AttributeError exception.

**Comment by rainwoodman** [comment]
Yes. I thought about the view() cast and realized it won't work. We need the `pm` object. so yes lets give it a try with `array_finalize` (though my impression is that you can never consistently cover all cases)

**Comment by nickhand** [comment]
Okay, do you want to take a stab at this, or do you want me to? 

Also, is there a reason you pass``buffer`` to ``add_attrs`` [here](https://github.com/rainwoodman/pmesh/blob/master/pmesh/pm.py#L50)?

**Comment by rainwoodman** [comment]
Please go ahead with this. I think it was kept there become didn't want to write that down twice. If you are to refactor it for array_finalize then you shall get rid of that.

**Comment by nickhand** [comment]
When you do view casting, it seems that there is no way to set the `base` attribute to a LocalBuffer from the python side. So, the `base` attribute ends up getting set as a numpy array rather than a LocalBuffer, and the returned object is actually useless for FFTs. Perhaps I'm missing something?

**Comment by rainwoodman** [comment]
It sounds very true, for we insist using memory allocated from pfft for pfft.

**Comment by nickhand** [comment]
consensus is that we will remove the subclass of ndarray to avoid these nasty issues

**Comment by rainwoodman** [comment]
I am trying to see how this affects nbodykit. Most of the pm.py API remains the same.

**Comment by nickhand** [comment]
from a quick look, I'm not sure it changes anything?

**Comment by nickhand** [comment]
There are a few problems, actually. Do you want me to push some fixes to master?

**Comment by rainwoodman** [comment]
I have been working on some fixes too. But lost track of it since last hour.

**Comment by rainwoodman** [comment]
OK I filed a PR at nbodykit that is intended to work with the new pmesh.

**Comment by nickhand** [comment]
closed via f7774a2 in v0.0.21
